### PR TITLE
Fix: storage migration shown on < API 30 & Amazon Builds

### LINKED ISSUE #13612: [BUG]: `allow_unsafe_migration` doesn't work as expected
### Checked for duplicates?

- [X] This issue is not a duplicate

### What are the steps to reproduce this bug?

Allow unsafe migration does not work as it should in all APIs reported by @Arthur-Milchior on Discord
```
## This file must *NOT* be checked into Version Control Systems,
# as it contains information specific to your local configuration.
#
# Location of the SDK. This is only used by Gradle.
# For customization when using a Version Control System, please read the
# header note.
#Sat Jul 30 19:00:38 CEST 2022
sdk.dir=/home/milchior/Android/Sdk
allow_unsafe_migration=true
```

Works for me on API27 but not on API30 or 32 and 33 for Arthur

### Expected behaviour

Migration button should be available 

### Actual behaviour

No migration button is shown even if its set to `true`

### Debug info

```text
alpha97-debug
```

### (Optional) Anything else you want to share?

_No response_

### Research

- [X] I am reporting a bug specific to AnkiDroid (Android app)
- [X] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue
- [ ] (Optional) I have confirmed the issue is not resolved in the latest alpha release ([instructions](https://docs.ankidroid.org/manual.html#betaTesting))

### LINKED ISSUE #13620: [BUG]: `NoSuchMethodError` for `java.io.File#toPath()` when trying to migrate on Android 6
### Checked for duplicates?

- [X] This issue is not a duplicate

### What are the steps to reproduce this bug?

* Open app
* Press migrate

### Expected behaviour

Migration starts

### Actual behaviour

App crashes with

```
   java.lang.NoSuchMethodError: No virtual method toPath()Ljava/nio/file/Path; in class Ljava/io/File; or its super classes (declaration of 'java.io.File' appears in /system/framework/core-libart.jar)
       at org.apache.commons.io.FileUtils.contentEquals(FileUtils.java:408)
       at com.ichi2.anki.servicelayer.scopedstorage.MigrateEssentialFiles.throwIfEssentialFilesAreMutated(MigrateEssentialFiles.kt:171)
       at com.ichi2.anki.servicelayer.scopedstorage.MigrateEssentialFiles.execute(MigrateEssentialFiles.kt:128)
       at com.ichi2.anki.servicelayer.scopedstorage.MigrateEssentialFiles$Companion$migrateEssentialFiles$2.invoke(MigrateEssentialFiles.kt:567)
       at com.ichi2.anki.servicelayer.scopedstorage.MigrateEssentialFiles$Companion$migrateEssentialFiles$2.invoke(MigrateEssentialFiles.kt:567)
       at com.ichi2.anki.exception.RetryableException$Companion.retryOnce(RetryableException.kt:32)
       at com.ichi2.anki.servicelayer.scopedstorage.MigrateEssentialFiles$Companion.migrateEssentialFiles$AnkiDroid_playDebug(MigrateEssentialFiles.kt:567)
       at com.ichi2.anki.servicelayer.scopedstorage.MigrateEssentialFiles$Companion.migrateEssentialFiles$AnkiDroid_playDebug$default(MigrateEssentialFiles.kt:529)
       at com.ichi2.anki.servicelayer.ScopedStorageService.migrateEssentialFiles(ScopedStorageService.kt:152)
       at com.ichi2.anki.DeckPicker$migrate$1$elapsedMillisDuringEssentialFilesMigration$1$1$1.invokeSuspend(DeckPicker.kt:2490)
       at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
       at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
       at kotlinx.coroutines.internal.LimitedDispatcher.run(LimitedDispatcher.kt:42)
       at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:95)
       at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:570)
       at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
       at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:677)
       at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:664)
```

### Debug info

```text
AnkiDroid Version = 2.16alpha97-debug

Android Version = 6.0.1

Manufacturer = motorola

Model = Moto G

Hardware = qcom

Webview User Agent = Mozilla/5.0 (Linux; Android 6.0.1; Moto G Build/MOB30Z; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/51.0.2704.106 Mobile Safari/537.36

ACRA UUID = 77e3b729-cfd0-475c-8f2e-d6cf913b1f4c

New schema = false

Scheduler = std2

Crash Reports Enabled = false

DatabaseV2 Enabled =
```

### (Optional) Anything else you want to share?

On ad538ed9efb2e8185032b8f55206654e465271b3

### Research

- [X] I am reporting a bug specific to AnkiDroid (Android app)
- [X] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue
- [ ] (Optional) I have confirmed the issue is not resolved in the latest alpha release ([instructions](https://docs.ankidroid.org/manual.html#betaTesting))

### PR DESCRIPTION
**Pull request by david-allison**
## Fixes
- Fixes #13612
- Fixes #13620

## Approach
We were using the condition `allFileAccessPermissionGranted` 

But this returns false before API 30  Instead define and use `collectionWillBeMadeInaccessibleAfterUninstall` which takes this condition into account  

## How Has This Been Tested?
* Unit Tests
* Tested on API 27 emulator

## Checklist
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/ScopedStorageService.kt [hunk 1]
```diff
@@ -308,16 +308,61 @@ object ScopedStorageService {
...
+     *    * The user installed with a newer Android version, and is more likely to expect this behavior.
+     * * The collection is now inaccessible
+     * * the user is on Android Q or below and Android **should** not revoke permissions
+     * * The user has the potential to grant [android.Manifest.permission.MANAGE_EXTERNAL_STORAGE]
+     * Returns `true` > Android 10 and the user has no way to access the collection on uninstall
+     * except for using another build of `com.ichi2.anki` pr manually copying files
+     * @see android.R.attr.preserveLegacyExternalStorage
+     * @see android.R.attr.requestLegacyExternalStorage
+     */
+    fun collectionWillBeMadeInaccessibleAfterUninstall(context: Context): Boolean {
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/ScopedStorageService.kt:358]: This is somewhat duplicated, but not enough to extract it (IMO). Review comments welcome

File: AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/ScopedStorageService.kt [hunk 2]
```diff
@@ -314,18 +314,55 @@ object ScopedStorageService {
...
+     * Whether the user's current collection is now inaccessible due to a 'reinstall'
+     *
+     * @return `false` if:
+     * * ⚠️ The directory will be **removed** on uninstall
+     *    * The user installed with a newer Android version, and is more likely to expect this behavior.
+     * * The collection is now inaccessible
+     * * the user is on Android Q or below and Android **should** not revoke permissions
+     * * The user has the potential to grant [android.Manifest.permission.MANAGE_EXTERNAL_STORAGE]
+     * Returns `true` > Android 10 and the user has no way to access the collection on uninstall
+     * except for using another build of `com.ichi2.anki` pr manually copying files
```
Comment (BrayanDSO) [inline AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/ScopedStorageService.kt:354]: "pr" at `pr manually...`

File: AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/ScopedStorageService.kt (hunk 1 above)
Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/ScopedStorageService.kt:358]: fine by me

**Review by BrayanDSO — CHANGES_REQUESTED** [review]

**Review by david-allison — COMMENTED** [review]
forgot to comment

**Review by BrayanDSO — APPROVED** [review]
LGTM

**Review by mikehardy — APPROVED** [review]
what a tangled web we weave

**Pull request merged**